fundamental-concepts/data-structures-stdlib/9_maps.md

```diff
@@ -0,0 +1,44 @@
+# Maps
+
+If you need to store associations from keys to values, the best data container to choose is a `Map`
+from the `DA.Next.Map` module. For example, we could have a map from phone numbers to contacts for
+quick number lookup:
+
+<pre class=file data-target="clipboard">
+import DA.Next.Map
+
+type PhoneBook =  Map PhoneNumber Person
+</pre>
+
+The `DA.Next.Map` module contains many useful functions to work with maps. Have a look at the
+[documentation](https://docs.daml.com/daml/stdlib/DA-Next-Map.html) and then take the quiz:
+
+>> Which function constructs a 'Map a b' from a list of tuples '[(a, b)]'? <<
+=== fromList
+
+>> Which functions constructs a 'Map a b' from a list of tuples '[(a, b)]', and combines duplicate keys in a given way? <<
+=== fromListWith
+
+>> How do you insert a new element "hello world" with key 3 in a map 'm: Map Int Text' ?<<
+=== insert 3 "hello world" m
+
+>> How do you lookup the element with key 3 in the map 'm' of the previous question? <<
+=== lookup 3 m
+
+>> What's the return type of the lookup of the previous question? <<
+() Text
+(*) Optional Text
+
+>> What function let's you check whether a given map is empty? <<<
+=== null
+
+>> Given the map 'm: Map Int Text', how do you filter for only even keys? <<
+(*) filterWithKey (\k _v ->  k `mod` 2 == 0) m
+() filter (k -> k `mod` 2 == 0) m
+
+>> How can you transform a given map 'm: Map a b' back to a list of tuples '[(a, b)]'? <<
+=== toList m
+
+>> What's the result of 'toList $ fromList [(1, "a"), (2, "b"), (2, "c"), (3, "d")]' ? <<
+=== [(1, "a"), (2, "c"), (3, "d")]
```

Comment:
```suggestion
>> Which function constructs a 'Map a b' from a list of tuples '[(a, b)]'? <<
=== fromList

>> Which functions constructs a 'Map a b' from a list of tuples '[(a, b)]', and combines duplicate keys in a given way? <<
=== fromListWith

>> How do you insert a new element "hello world" with key 3 in a map 'm: Map Int Text' ?<<
=== insert 3 "hello world" m

<!--
>> How do you lookup the element with key 3 in the map 'm' of the previous question? <<
=== lookup 3 m

>> What's the return type of the lookup of the previous question? <<
() Text
(*) Optional Text
-->
>> What function let's you check whether a given map is empty? <<<
=== null
<!--
>> Given the map 'm: Map Int Text', how do you filter for only even keys? <<
(*) filterWithKey (\k _v ->  k `mod` 2 == 0) m
() filter (k -> k `mod` 2 == 0) m

>> How can you transform a given map 'm: Map a b' back to a list of tuples '[(a, b)]'? <<
=== toList m

>> What's the result of 'toList $ fromList [(1, "a"), (2, "b"), (2, "c"), (3, "d")]' ? <<
=== [(1, "a"), (2, "c"), (3, "d")]
-->
```

The number of questions was overwhelming. Three would be best, four max.

Reply:
a few have been removed, again, all solutions are at the bottom, so you can easily skip them.